avocado/plugins/spawners/remote.py

```diff
@@ -0,0 +1,210 @@
+import asyncio
+import contextlib
+import json
+import logging
+import os
+import shlex
+
+from aexpect import remote
+
+from avocado.core.plugin_interfaces import Init, Spawner
+from avocado.core.settings import settings
+from avocado.core.spawners.common import SpawnerMixin, SpawnMethod
+
+LOG = logging.getLogger(__name__)
+
+
+class RemoteSpawnerException(Exception):
+    """Errors more closely related to the spawner functionality"""
+
+
+class RemoteSpawnerInit(Init):
+
+    description = "Remote (host) based spawner initialization"
+
+    def initialize(self):
+        section = "spawner.remote"
+
+        help_msg = "List of already available remote host slots to spawn in"
+        settings.register_option(
+            section=section, key="slots", help_msg=help_msg, key_type=list, default=[]
+        )
+
+        help_msg = "Remote host setup hook command to customize optional new hosts"
+        settings.register_option(
+            section=section, key="setup_hook", help_msg=help_msg, default=""
+        )
+
+
+def with_slot_reservation(fn):
+    """
+    Decorator for slot cache context manager.
+
+    :param fn: function to run with slot reservation
+    :type fn: function
+    :returns: same function with the slot now reserved
+    :rtype: function
+
+    The main reason for the decorator is to not have to indent the entire
+    task running function in order to safely release the slot upon any error.
+    """
+
+    async def wrapper(self, runtime_task):
+        with RemoteSpawner.reserve_slot(self, runtime_task) as slot:
+            runtime_task.spawner_handle = slot
+            return await fn(self, runtime_task)
+
+    return wrapper
+
+
+class RemoteSpawner(Spawner, SpawnerMixin):
+
+    description = "Remote (host) based spawner"
+    METHODS = [SpawnMethod.STANDALONE_EXECUTABLE]
+    slots_cache = {}
+
+    @staticmethod
+    async def run_remote_cmd_async(session, command, timeout):
+        loop = asyncio.get_event_loop()
+        try:
+            status, output = await loop.run_in_executor(
+                None, session.cmd_status_output, command, timeout
+            )
+        except exceptions.ShellTimeoutError:
+            status, output = 2, f"Remote command timeout of {timeout} reached"
+        except exceptions.ShellProcessTerminatedError:
+            status, output = 2, f"Remote command terminated prematurely"
+        return status, output
+
+    @contextlib.contextmanager
+    def reserve_slot(self, runtime_task):
```

Comment:
This is also pretty much identical to the lxc version... It makes an even stronger point towards sharing more of this code IMO.